### PR TITLE
Removed the define for the `JETPACK_AUTOLOAD_DEV` constant

### PR DESCRIPTION
**Pull request by ObliviousHarmony**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Originally this define was set to avoid the potential confusion that could be created when it is not set. After a chat with @nerrad we decided that it should be set in the developer's `wp-config.php` so that WooCommerce Core doesn't have any unintended side-effects for other plugins that use the autoloader.

We can address the confusion in our feature plugins by giving a notice when using a developer build without the constant defined as `true`.

### Special Notes

Once this PR is merged the `woocommerce-admin` and `woo-gutenberg-products-block` plugins will experience some strange behavior. Developers working on the `woo-gutenberg-products-block` plugin will be fine but those working on `woocommerce-admin` will need to add the `JETPACK_AUTOLOAD_DEV` constant to their `wp-config.php`.

### How to test the changes in this Pull Request:

1. Clone `woocommerce-admin` into your plugins and make sure the resources are built and the plugin activated.
2. Add a logging statement to the package's `Package.php` file's `init()` method.
3. Go to the WooCommerce Dashboard and then check your log. Without the PR you'll see the log message but with it you will not.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Fix - Removed the `JETPACK_AUTOLOAD_DEV` define

### CODE REVIEW

**Review by Konamiman — APPROVED** [review]

**Review by peterfabian — COMMENTED** [review]
I tried the following scenarios:

| WC Blocks | JETPACK_AUTOLOAD_DEV | code used |
| ------------- | ------------- | ------------- |
| packaged | false | package |
| packaged + git cloned `main` and built | true | git cloned |
| packaged + plugin | true | plugin |
| packaged + plugin | false | package |

I'm curious about the last row in the table ^. 
Since we wanted to use plugins to enable new features for customers before they are activated in WC core, would that still work? Or are the new upcoming versions of Blocks and WC Admin going to solve this from the side of the plugin?

**Comment by peterfabian** [comment]
Tests For WC Admin:

| WC Admin | JETPACK_AUTOLOAD_DEV | code used |
| ------------- | ------------- | ------------- |
| packaged | false | package |
| packaged + git cloned `main` and built | n/a | fatal error± |
| packaged + git cloned `tag/1.3.2` and built | true | plugin |
| packaged + git cloned `tag/1.3.2` and built | false | plugin |
| packaged + plugin | true | plugin |
| packaged + plugin | false | plugin |

 think the behavior of WC Admin and Blocks should ideally be the same, no?

±: the error is as follows: 
```
PHP Fatal error:  require_once(): Failed opening required '/srv/www/wordpress-one/public_html/wp-content/plugins/woocommerce-admin//includes/feature-config.php' (include_path='.:/usr/share/php') in /srv/www/wordpress-one/public_html/wp-content/plugins/woocommerce-admin/src/FeaturePlugin.php on line 75
[03-Aug-2020 13:11:56 UTC] PHP Stack trace:
[03-Aug-2020 13:11:56 UTC] PHP   1. {main}() /srv/www/wordpress-one/public_html/wp-admin/admin-ajax.php:0
[03-Aug-2020 13:11:56 UTC] PHP   2. require_once() /srv/www/wordpress-one/public_html/wp-admin/admin-ajax.php:22
[03-Aug-2020 13:11:56 UTC] PHP   3. require_once() /srv/www/wordpress-one/public_html/wp-load.php:37
[03-Aug-2020 13:11:56 UTC] PHP   4. require_once() /srv/www/wordpress-one/public_html/wp-config.php:87
[03-Aug-2020 13:11:56 UTC] PHP   5. include_once() /srv/www/wordpress-one/public_html/wp-settings.php:377
[03-Aug-2020 13:11:56 UTC] PHP   6. Automattic\WooCommerce\Admin\FeaturePlugin->init() /srv/www/wordpress-one/public_html/wp-content/plugins/woocommerce-admin/woocommerce-admin.php:69

```

**Review by peterfabian — APPROVED** [review]
As noted by @ObliviousHarmony , the problem I mentioned will be solved by updating composer versions in subpackages--they won't be considered development versions anymore. Thus it should be fine to merge this. That being said, it would be nice to harmonize our packages, as each one uses autoloader differently.

**Comment by peterfabian** [comment]
Added label `high impact` as it needs to be documented well for the community, developers and HEs.

**Comment by nerrad** [comment]
> I'm curious about the last row in the table ^.
> Since we wanted to use plugins to enable new features for customers before they are activated in WC core, would that still work? Or are the new upcoming versions of Blocks and WC Admin going to solve this from the side of the plugin?

I think you already have this clarified, but as a point of reference, Woo Blocks [recently merged some changes to the build process](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2949) so that future releases of Woo Blocks (beginning with `3.2.0`) will have the proper package version set in `composer.json` before the autoloader is generated. That will ensure user's of the feature plugin will have plugin classes loaded _if the version of the plugin is higher than what's found in WooCommerce core_.
